torch/functional.py

```diff
@@ -1589,3 +1590,49 @@ def _lu_no_infos(A, pivot=True, get_infos=False, out=None):
 
 def align_tensors(*tensors):
     raise RuntimeError('`align_tensors` not yet implemented.')
+
+def result_type(*arrays_and_dtypes: Union[Tensor, torch.dtype, bool, int, float, complex]) -> torch.dtype:
+    """result_type(*arrays_and_dtypes: Union[Tensor, dtype, bool, int, float, complex]) -> dtype
+
+Returns the :class:`torch.dtype` that would result from performing an arithmetic
```

Comment:
This statement is a little misleading now since the inputs don't have to be exclusively tensors, but that's not a significant issue. I do have a question about the prominence of dtypes, however. When including a dtype, is that like including a scalar with that dtype, a 0D tensor with that dtype, or a 1+D tensor with that dtype in the arithmetic operation?

This distinction is important because PyTorch currently treats these things differently:

```
# 1+D tensor dtypes have a higher priority than 0D tensor dtypes
a = torch.tensor((1.0,), dtype=torch.float16)
b = torch.tensor(1.0)
a + b
: tensor([2.], dtype=torch.float16)
```

Things get even weirder for when type promotion results in a new dtype:

```
# 1+D tensor dtypes have higher priority than 0D tensor dtypes
a = torch.tensor((1.0,), dtype=torch.float16)
b = torch.tensor(1.0, dtype=torch.bfloat16)
a + b
: tensor([2.], dtype=torch.float16)
: 

# When both tensors are 1+D type promotion produces a float32 tensor!
a = torch.tensor((1.0,), dtype=torch.float16)
b = torch.tensor((1.0,), dtype=torch.bfloat16)
a + b
: tensor([2.])

# When both tensors are 0D type promotion also produces a float32 tensor
a = torch.tensor(1.0, dtype=torch.float16)
b = torch.tensor(1.0, dtype=torch.bfloat16)
a + b
: tensor(2.)
```

Basically PyTorch says: "What's the highest type kind of any operand?" Where the type kinds are ordered bool int float complex. Then, among the objects with the highest type kind, is there a 1+D tensor? If so the 1+D tensors with the highest type kind define the result type. If there are no 1+D tensors then 0D tensors are reviewed the same way, and finally scalars are assigned the default dtype for the type kind and evaluated (the default dtypes for each kind are bool, long, float32, complex64, although a user can set the float and complex defaults to float64 and complex128). 

So when a user queries using dtype, which of the above priorities is it given? 

Note that if we change type promotion to conform with NumPy, JAX, and the Python Array API then this picture would be simplified because those frameworks/specifications don't treat 0D tensors differently than 1+D tensors.

Reply:
I think if it takes this long to explain the type promotion rules, then they are probably too complex. From the array api standard... 

> Accordingly, zero-dimensional arrays must be subject to the same type promotion rules as dimensional arrays.

and for mixing arrays and python scalars

> Convert the scalar to a 0-D array with the same dtype as that of the array used in the expression.

I think these rules make sense because if we want to multiply all elements of a tensor `x` by 2, we don't want to be thinking about what dtype the tensor is and instead just write `x * 2` and have the result be the same dtype as the input tensor.

Reply:
@mruberry @heitorschueroff 
> So when a user queries using dtype, which of the above priorities is it given?

Dtype inputs are handled just like as 1+D Tensors.

torch.result_type of the current master branch distinguishes inputs of 1+-D Tensors, 0-D Tensors, and scalars respectively. Currently, this PR does not break the backward compatibility of this behavior.

Reply:
Thanks for the analysis, @asi1024